neurodsp/sim/params.py

```diff
@@ -0,0 +1,589 @@
+        """Register an iterator definition.
+
+        Parameters
+        ----------
+        name : str
+            Name to give the registered iterator.
+        label : str
+            Label for the simulation parameters.
+        update : str
+            Name of the parameter to update.
+        values : 1d array
+            Values to iterate across.
+        component : str, optional
+            Which component to update the parameter in.
+            Only used if the parameter definition is for a multi-component simulation.
+        """
+
+        self._iters[name] = {
+            'label' : label,
+            'update' : update,
+            'values' : values,
+            'component' : component,
+        }
+
+
+    def register_group_iters(self, group, clear=False):
+        """Register a group of simulation iterators.
+
+        Parameters
+        ----------
+        group : list of list or list of dict
+            Set of simulation iterator definitions.
+        clear : bool, optional, default: False
+            If True, clears current parameter iterators before adding new group.
+        """
+
+        if clear:
+            self.clear()
+
+        for iterdef in group:
+            if isinstance(iterdef, list):
+                self.register_iter(*iterdef)
+            elif isinstance(iterdef, dict):
+                self.register_iter(**iterdef)
+
+
+    def update_iter(self, label, attribute, new_value):
+        """Update the definition of an iterator.
+
+        Parameters
+        ----------
+        label : str
+            Label for the iterator.
+        attribute : str
+            Name of the attribute to update.
+        new_value : obj
+            New value to give the attribute.
+        """
+
+        self._iters[label][attribute] = new_value
+
+
+    def clear(self, clear_iters=True, clear_params=False, clear_base=False):
+        """"Clear iterator and/or parameter definitions.
+
+        Parameters
+        ----------
+        clear_iters : bool, optional, default: True
+            Whether to clear the currently defined iterators.
+        clear_params : bool, optional, default: False
+            Whether to clear the currently defined simulation parameters.
+        clear_base : bool, optional, default: False
+            Whether to also clear base parameters.
+            Only applied if `clear_params` is True.
+        """
+
+        if clear_iters:
+            self._iters = {}
+
+        if clear_params:
+            super().clear(clear_base=clear_base)
+
+
+class SimSamplers(SimParams):
+    """Object for sampling simulation parameter definitions.
+
+    Parameters
+    ----------
+    n_seconds : float
+        Simulation time, in seconds.
+    fs : float
+        Sampling rate of simulated signal, in Hz.
+    n_samples : int, optional
+        The number of parameter iterations to set as max.
+        If None, samplers are created as infinite generators.
+    sim_params : SimParams
```

Comment:
Same comment as left on SimIters about having a method on SimParams that returns an initialized SimSampler, e.g.:

```
sim_sampler = sim_params.to_sampler(n_samples)
sim_iters = sim_params.to_iter()
```

This may reduce complexity of the either/or in the init here.

Reply:
Good catch / note - it works really nicely to have these "export" methods defined on SimParams that one can use if you already have a SimParams object, and then the derived objects (SimIters, SimSamplers) no longer optionally accept a SimParams object at init, and have a cleaner initialization! Updated!